### PR TITLE
Fetch queues

### LINKED ISSUE #1210: validation queue clean callback map
# Overview / Motivation
need to delete callbacks from validation queue after they are run 

# The Task
TODO: Clearly describe the issue requirements here...

# Implementation Notes
TODO: Add links to relevant resources, specs, related issues, etc...

# Contribution Guidelines
Important: Issue assignment to developers will be by the order of their application and proficiency level according to the tasks complexity. We will not assign tasks to developers who have'nt introduced themselves on our Gitter [dev channel](https://gitter.im/spacemesh-os/Lobby)

1. Introduce yourself on go-spacemesh [dev chat channel](https://gitter.im/spacemesh-os/Lobby) - ask our team any question you may have about this task
2. Fork branch `develop` to your own repo and work in your repo
3. You must document all methods, enums and types with [godoc comments](https://blog.golang.org/godoc-documenting-go-code)
4. You must write go unit tests for all types and methods when submitting a component, and integration tests if you submit a feature
5. When ready for code review, submit a PR from your repo back to branch `develop`
6. Attach relevant issue to PR

### PR DESCRIPTION
**Pull request by almogdepaz**
added fetchQueue

added txQueue

added atxQueue

added blockQueue

syncer now uses queues to fetch

unify ShortString() and ShortId()

remove duplicated code from sync pkg

add log interface to log pkg

add item interface for txs atxs and blocks

use hash32 as id for txs atxs and blocks

fix #1210

### CODE REVIEW

File: common/types/activation.go [hunk 1]
```diff
@@ -130,26 +136,26 @@ func NewActivationTxWithChallenge(poetChallenge NIPSTChallenge, coinbase Address
...
 
 func (atxh *ActivationTxHeader) TargetEpoch(layersPerEpoch uint16) EpochId {
 	return atxh.PubLayerIdx.GetEpoch(layersPerEpoch) + 1
 }
 
-func (atxh *ActivationTxHeader) SetId(id *AtxId) {
-	atxh.id = id
+func (atxh *ActivationTxHeader) SetId(id AtxId) {
+	if atxh.AtxId != *EmptyAtxId {
+		return
```
Comment (gavraz) [inline common/types/activation.go]: What does the flow of an already set atxid means?
Shouldn't it be an error? a panic? should we report back to the caller?

File: common/types/block.go [hunk 2]
```diff
@@ -74,6 +83,12 @@ type MiniBlock struct {
 	AtxIds []AtxId
 }
 
+func (t BlockID) AsHash32() Hash32 {
```
Comment (gavraz) [inline common/types/block.go:86]: Rename to Hash32

File: mesh/mesh.go [hunk 3]
```diff
@@ -279,7 +279,7 @@ func (m *Mesh) GetLatestView() []types.BlockID {
 	}
 	view := make([]types.BlockID, 0, len(layer.Blocks()))
 	for _, blk := range layer.Blocks() {
-		view = append(view, blk.Id)
+		view = append(view, blk.ID())
```
Comment (gavraz) [inline mesh/mesh.go:282]: Rename to Id() ?

File: sync/fetch_queue.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+		if len(bjb.ids) == 0 {
+			return fmt.Errorf("channel closed")
+		}
+
+		if len(bjb.items) == 0 {
+			fq.Warning("could not fetch any items ") //todo print ids
+			continue
+		}
+
+		fq.handle(out.(fetchJob))
```
Comment (gavraz) [inline sync/fetch_queue.go]: Not sure about performance but if you already made the cast just keep it and use it (bjb).

File: sync/fetch_queue.go [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+	deps := fq.pending[id]
+	delete(fq.pending, id)
+	fq.Unlock()
+
+	for _, dep := range deps {
+		dep <- valid
+	}
+}
+
+//returns txs out of txids that are not in the local database
```
Comment (gavraz) [inline sync/fetch_queue.go]: Fix comment, to my understanding it is not only txs we handle here

File: sync/fetch_queue.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+			continue
+		} else {
+			return nil, errors.New(fmt.Sprintf("item %v was not found after fetch was done", hex.EncodeToString(id[:])))
+		}
+	}
+	return txs, nil
+}
+
+//todo make the queue generic
+type txQueue struct {
```
Comment (gavraz) [inline sync/fetch_queue.go:143]: It is possible that you can avoid txQueue (and atxQueue).
1. fetchQueue can recv items that implements the func Hash32 (then, you don't need the cast in HandleTxs for example)
2. The fetchQueue.handle function can be constructed, with its dependencies, prior to the construction of the fetchQueue and passed to it. Then, the handle function signature can be simple and type-free as its dependencies have already been injected to it.

Look into it to see if it is possible as it can unburden the need in extra structs.

File: sync/fetch_queue.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+					continue
+				}
+			}
+			invalidate(id, false)
+		}
+	}
+
+}
+
+func getDoneChan(deps []chan bool) chan bool {
```
Comment (gavraz) [inline sync/fetch_queue.go:283]: This feels like a wg logic. See if you can simplify it to use wg.

File: sync/fetch_queue.go [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+func (fq *fetchQueue) addToPendingGetCh(ids []types.Hash32) chan bool {
+	return getDoneChan(fq.addToPending(ids))
+}
+
+func (fq *fetchQueue) addToPending(ids []types.Hash32) []chan bool {
+	fq.Lock()
+	deps := make([]chan bool, 0, len(ids))
+	var idsToAdd []types.Hash32
+	for _, id := range ids {
+		ch := make(chan bool, 1)
```
Comment (gavraz) [inline sync/fetch_queue.go:93]: Note that you are creating at least one channel for each id. If you have duplicated requests you will have multiple channels per id.
I think you should think of an alternative (more simple) mechanism or simplify it by having only one channel per item and record the number of "subscribers" per item so you can notify all of them (without requiring creating a channel to each one of them). Think of buffer limitations as it may suggest that this "simple" solution is not feasible.
Look into the option of having a wg per item collection (@see my comment on getDoneChan).

File: sync/syncer.go [hunk 9]
```diff
@@ -287,9 +288,29 @@ func (s *Syncer) Synchronise() {
 	s.handleNotSynced(currentSyncLayer)
 }
 
+func (s *Syncer) FastValidation(block *types.Block) error {
```
Comment (gavraz) [inline sync/syncer.go]: Why is it a method of syncer? why not inject it in the constuctor?

File: sync/syncer.go [hunk 10]
```diff
@@ -433,43 +442,54 @@ func validateUniqueTxAtx(b *types.Block) error {
 	return nil
 }
 
-func (s *Syncer) BlockSyntacticValidation(block *types.Block) ([]*types.AddressableSignedTransaction, []*types.ActivationTx, error) {
-	// run block fast validation
-	if err := s.blockFastValidator(block); err != nil {
-		s.With().Error("BlockSyntacticValidation: block validation failed", log.BlockId(uint64(block.ID())), log.Err(err))
+func (s *Syncer) blockSyntacticValidation(block *types.Block) ([]*types.AddressableSignedTransaction, []*types.ActivationTx, error) {
```
Comment (gavraz) [inline sync/syncer.go:445]: Are you rebased on develop? Where is the call to fastValidation?

File: common/types/block.go (hunk 2 above)
Comment (almogdepaz) [reply common/types/block.go:86]: its on purpose usually Hash32 refers to the hash of the object 
here its more a conversion

File: mesh/mesh.go (hunk 3 above)
Comment (almogdepaz) [reply mesh/mesh.go:282]: cool but lets do it in a different pr because it involves lots of modules that are not here

File: sync/fetch_queue.go (hunk 6 above)
Comment (almogdepaz) [reply sync/fetch_queue.go:143]: because the invalidate method is dependent on the queue instance it creates more mess than it saves right now
but after we unify the ids this will all disappear

File: sync/syncer.go (hunk 9 above)
Comment (almogdepaz) [reply sync/syncer.go]: i think its more readable this way syncer fits into the ValidationInfra interface that holds all methods for validation queue 
seams cleaner to me, it should also be private ill change it

File: sync/syncer.go (hunk 10 above)
Comment (almogdepaz) [reply sync/syncer.go]: it happens inside the validation queue for blocks that are fetched
and in the block_listener for blocks that are gossiped

File: sync/fetch_queue.go (hunk 8 above)
Comment (almogdepaz) [reply sync/fetch_queue.go:93]: i replayed above

File: sync/fetch_queue.go (hunk 7 above)
Comment (almogdepaz) [reply sync/fetch_queue.go:283]: basically i was thinking the same thing, the if i could use chan struct{} or something it would be much simpler but since i need a bool for each id replacing this will probably include a new struct with a mutex and a bool field,

ill open a task to check this possibility,  right now this was the easiest implementation and returning a chan for subscribing  is a pretty common pattern 
#1398

File: sync/fetch_queue.go (hunk 6 above)
Comment (gavraz) [reply sync/fetch_queue.go:143]: Ok cool. Either way, it is cleaner than before.

File: common/types/activation.go (hunk 1 above)
Comment (almogdepaz) [reply common/types/activation.go]: added an error, this can happen on atxs that we fetch

File: sync/fetch_queue.go (hunk 7 above)
Comment (gavraz) [reply sync/fetch_queue.go:283]: Ok, I see. The bool indeed complicates it a little bit. But please open an issue it at least worth another look.

File: sync/syncer.go (hunk 9 above)
Comment (gavraz) [reply sync/syncer.go]: Please rethink it. Having a method memeber gives you two main adbvantages:
1. Easier to test (just replace the member func with whatever mock you like)
2. It unburdens the syncer from this extra knowledge. It is pretty simple to say "I want to validate a block so I have a validation function" than "I want to validate a block so I keep all the components required so I can use them when I need to validate".

Regarding the validation infra - there is no wrong with having an internal struct for it (if the interface really requires just validation) or simply passing the validationFunc as before. IMO, You should try to avoid adding extra knowledge/abilities to the syncer and break it to (interface) components that tend to be smaller and simpler.

Comment (almogdepaz) [reply sync/syncer.go]: its not that i disagree with your points 
its just that im aiming for this to be changed soon and building on top of it will make it harder in the next pr :)

File: common/types/activation.go [hunk 11]
```diff
@@ -24,19 +25,25 @@ func (l EpochId) FirstLayer(layersPerEpoch uint16) LayerID {
 	return LayerID(uint64(l) * uint64(layersPerEpoch))
 }
 
-type AtxId struct {
-	Hash32
+type AtxId Hash32
+
+func (t AtxId) ShortString() string {
```
Comment (zalmen) [inline common/types/activation.go]: shouldn't it be a pointer receiver? `ShortString` is called quite often, on every relevant log print.

File: common/types/activation.go [hunk 12]
```diff
@@ -24,19 +25,25 @@ func (l EpochId) FirstLayer(layersPerEpoch uint16) LayerID {
...
+	return t.Hash32().ShortString()
+}
+
+func (t AtxId) Hash32() Hash32 {
```
Comment (zalmen) [inline common/types/activation.go]: is it better than casting?

File: common/types/activation.go [hunk 13]
```diff
@@ -130,26 +137,28 @@ func NewActivationTxWithChallenge(poetChallenge NIPSTChallenge, coinbase Address
 }
 
 func (atxh *ActivationTxHeader) Id() AtxId {
-	if atxh.id == nil {
-		panic("id field must be set")
+	if atxh.AtxId == *EmptyAtxId {
```
Comment (zalmen) [inline common/types/activation.go]: `AtxId` is always set when doing unmarsheling/creating new atx, there is no case where you hold an atx struct with un-initialized id

File: common/types/block.go [hunk 14]
```diff
@@ -10,7 +10,16 @@ import (
 )
 
 type BlockID uint64
-type TransactionId [32]byte
+type TransactionId Hash32
+
+func (t TransactionId) ShortString() string {
```
Comment (zalmen) [inline common/types/block.go]: pointer receiver?

File: sync/block_listener.go [hunk 15]
```diff
@@ -70,38 +62,38 @@ func (bl *BlockListener) ListenToGossipBlocks() {
...
-				bl.Error("got empty message while listening to gossip blocks")
-				return
-			}
-			var blk types.Block
-			err := types.BytesToInterface(data.Bytes(), &blk)
-			if err != nil {
-				bl.Error("received invalid block %v", data.Bytes(), err)
-				return
-			}
+			go func() {
```
Comment (zalmen) [inline sync/block_listener.go:64]: Are the todos above still relevant?

File: sync/block_listener.go [hunk 16]
```diff
@@ -70,38 +62,38 @@ func (bl *BlockListener) ListenToGossipBlocks() {
...
+					return
+				}
 
-			if bl.HandleNewBlock(&blk) {
-				data.ReportValidation(config.NewBlockProtocol)
-			}
-			bl.wg.Done()
-			//}()
+				if bl.HandleNewBlock(&blk) {
+					data.ReportValidation(config.NewBlockProtocol)
```
Comment (zalmen) [inline sync/block_listener.go]: As we talked f2f, move the `ReportValidation` into `HandleNewBlock` and call it after the validation

File: sync/fetch_queue.go [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+		}
+
+		bjb, ok := out.(fetchJob)
+		if !ok {
+			fq.Error(fmt.Sprintf("Type conversion err %v", out))
+			continue
+		}
+
+		if len(bjb.ids) == 0 {
+			return fmt.Errorf("channel closed")
```
Comment (zalmen) [inline sync/fetch_queue.go:68]: Why is this the error for this case? Which channel is closed?

File: sync/fetch_queue.go [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+		}
+
+		if len(bjb.items) == 0 {
+			fq.Warning("could not fetch any items ") //todo print ids
```
Comment (zalmen) [inline sync/fetch_queue.go]: consider implementing the todo, it'll ease debugging

File: sync/fetch_queue.go [hunk 19]
```diff
@@ -0,0 +1,315 @@
...
+	delete(fq.pending, id)
+	fq.Unlock()
+
+	for _, dep := range deps {
+		dep <- valid
+	}
+}
+
+//returns items out of itemIds that are not in the local database
+func (tq *fetchQueue) Handle(itemIds []types.Hash32) ([]Item, error) {
```
Comment (zalmen) [inline sync/fetch_queue.go]: Why is it public? It seems that it is called only from within the package

File: sync/fetch_queue.go [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+	if len(itemIds) == 0 {
+		tq.Debug("handle empty item ids slice")
+		return nil, nil
```
Comment (zalmen) [inline sync/fetch_queue.go:121]: amm are you sure it is safe to return a nil for `[]item`? Shouldn't you return an empty slice?

File: sync/fetch_queue.go (hunk 19 above)
Comment (zalmen) [reply sync/fetch_queue.go]: Also, since it has a rather non-verbose name, please add more detailed documentation describing what the method does

File: sync/fetch_queue.go [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+			return nil, errors.New(fmt.Sprintf("could not fetch all items"))
+		}
+
+		unprocessedItems, processedItems, missing = tq.checkLocal(itemIds)
+		if len(missing) > 0 {
+			return nil, errors.New("could not find all items even though fetch was successful")
+		}
+	}
+
+	txs := make([]Item, 0, len(itemIds))
```
Comment (zalmen) [inline sync/fetch_queue.go]: IMO you should remove this part, it is too defensive

Comment (zalmen) [reply sync/fetch_queue.go]: Also, it is out of scope for the fetch queue

File: sync/fetch_queue.go [hunk 22]
```diff
@@ -0,0 +1,315 @@
...
+			}
+		}
+		doneChan <- alldone
+		close(doneChan)
+	}()
+	return doneChan
+}
+
+//todo get rid of this, send proofs with atxs
+func fetchProofCalcId(fetchPoetProof FetchPoetProofFunc, fj fetchJob) {
```
Comment (zalmen) [inline sync/fetch_queue.go:301]: it will be nice if you'll use the fetch Q for fetching the PoET as well

File: sync/handler.go [hunk 23]
```diff
@@ -10,8 +11,8 @@ import (
 
 func newLayerHashRequestHandler(layers *mesh.Mesh, logger log.Log) func(msg []byte) []byte {
 	return func(msg []byte) []byte {
-		logger.Debug("handle layer hash request")
 		lyrid := util.BytesToUint64(msg)
+		logger.Info("handle layer %v hash request", lyrid)
```
Comment (zalmen) [inline sync/handler.go]: use `With()`

File: sync/handler.go [hunk 24]
```diff
@@ -50,24 +51,33 @@ func newLayerBlockIdsRequestHandler(layers *mesh.Mesh, logger log.Log) func(msg
...
 
-		bbytes, err := types.InterfaceToBytes(blk)
+		var blocks []types.Block
+		for _, bid := range blockids {
+			var id = util.BytesToUint64(bid.Bytes())
+			logger.Debug("handle block %v request", id)
+			blk, err := msh.GetBlock(types.BlockID(id))
+			if err != nil {
+				logger.Error("Error handling block request message, with BlockID: %d and err: %v", bid, err)
+				return nil
```
Comment (zalmen) [inline sync/handler.go]: why return? If there is a problem with one block, why don't you continue working on the rest of the blocks? It is ok to return a partial response

File: sync/fetch_queue.go (hunk 22 above)
Comment (almogdepaz) [reply sync/fetch_queue.go:301]: agreed, does it matter if we do it after the "ids"?
poetproofs ids need to be also hash32 if we want to use the fetchQueue

File: sync/fetch_queue.go (hunk 20 above)
Comment (almogdepaz) [reply sync/fetch_queue.go:121]: why not ? because it's an interface ?

File: common/types/activation.go (hunk 13 above)
Comment (almogdepaz) [reply common/types/activation.go]: its not private anymore because it caused me lots of pain lets fix it in another pr with the rest of the ids

File: common/types/activation.go (hunk 12 above)
Comment (almogdepaz) [reply common/types/activation.go]: type conversion   Hash32(t)
type assertion     t.(Hash32)
its not that the one is better than the other
assertion can be used only on interfaces

File: sync/fetch_queue.go (hunk 17 above)
Comment (almogdepaz) [reply sync/fetch_queue.go:68]: the main output channel this means we cant continue with this instance 
i can do a panic but i thought id leave the option to the caller

File: sync/requests.go [hunk 25]
```diff
@@ -3,46 +3,50 @@ package sync
...
-				s.Info("received empty layer ids response")
+			if len(msg) == 0 || msg == nil {
+				s.Warning("peer %v responded with nil to layer %v request", peer, lyr)
 				return
 			}
-
 			ids, err := types.BytesToBlockIds(msg)
 			if err != nil {
-				s.Error("could not unmarshal layer ids response: %v", err)
+				s.Error("could not unmarshal mesh.LayerIDs response")
```
Comment (zalmen) [inline sync/requests.go]: Why not print the err?

File: sync/requests.go [hunk 26]
```diff
@@ -3,46 +3,50 @@ package sync
...
+		foo := func(msg []byte) {
 			defer close(ch)
-			if msg == nil {
-				s.Info("received empty layer ids response")
+			if len(msg) == 0 || msg == nil {
+				s.Warning("peer %v responded with nil to layer %v request", peer, lyr)
 				return
 			}
-
 			ids, err := types.BytesToBlockIds(msg)
```
Comment (zalmen) [inline sync/requests.go:29]: Are you validating somewhere that the response matches the request?

File: sync/requests.go [hunk 27]
```diff
@@ -263,3 +194,28 @@ func validatePoetRef(proofMessage types.PoetProofMessage, poetProofRef []byte) (
 
 	return true, nil
 }
+
+func validateItemIds(ids []types.Hash32, items []Item) (bool, error) {
```
Comment (zalmen) [inline sync/requests.go:198]: you have exactly the same logic in fetch_queue.go, consider extracting it to a common method for better code re-use

File: sync/syncer.go [hunk 28]
```diff
@@ -287,9 +288,29 @@ func (s *Syncer) Synchronise() {
 	s.handleNotSynced(currentSyncLayer)
 }
 
+func (s *Syncer) fastValidation(block *types.Block) error {
```
Comment (zalmen) [inline sync/syncer.go:291]: `fastValidation` is a misleading name. `BlockSignedAndEligible` is rather slow since it validates the BLS signature

File: sync/syncer.go [hunk 29]
```diff
@@ -488,57 +504,37 @@ func (s *Syncer) validateVotes(blk *types.Block) bool {
...
-func (s *Syncer) DataAvailability(blk *types.Block) ([]*types.AddressableSignedTransaction, error, []*types.ActivationTx, error) {
-	var txs []*types.AddressableSignedTransaction
+func (s *Syncer) DataAvailabilty(blk *types.Block) ([]*types.AddressableSignedTransaction, []*types.ActivationTx, error) {
+
+	var txres []*types.AddressableSignedTransaction
 	var txerr error
-	wg := sync.WaitGroup{}
-	wg.Add(2)
-
-	go func() {
```
Comment (zalmen) [inline sync/syncer.go:511]: Why did you change it to be sequential? What is wrong with validating the Txs and the Atxs concurrently?

File: sync/validation_queue.go [hunk 30]
```diff
@@ -4,34 +4,57 @@ import (
...
+	callbacks     map[interface{}]func(res bool) error
+	depMap        map[interface{}]map[types.BlockID]struct{}
+	reverseDepMap map[types.BlockID][]interface{}
+	visited       map[types.BlockID]struct{}
+}
+
+func NewValidationQueue(srvr WorkerInfra, conf Configuration, msh ValidationInfra, checkLocal CheckLocalFunc, lg log.Log) *blockQueue {
+	vq := &blockQueue{
+		fetchQueue: fetchQueue{
+			Log:                 srvr.WithName("atxFetchQueue"),
```
Comment (zalmen) [inline sync/validation_queue.go]: "atxFetchQueue" ?

File: sync/validation_queue.go [hunk 31]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
-	if vq.addDependencies(blk, s.GetBlock) == false {
-		return nil
+func (vq *blockQueue) handleBlock(bjb fetchJob) {
+	//todo movem to batch block request
+	bid := types.BlockID(util.BytesToUint64(bjb.ids[0].Bytes()))
+	if bjb.items == nil {
+		//no items fetched
+		vq.updateDependencies(bid, false)
+		vq.Error(fmt.Sprintf("could not retrieve a block in view "))
+		//return
```
Comment (zalmen) [inline sync/validation_queue.go]: You must return here, right? Otherwise, you'll get a pointer violation in the next few lines..

File: sync/validation_queue.go [hunk 32]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
 	}
-
-	vq.callbacks[blk.ID()] = func() error {
-		return nil
+	block := bjb.items[0].(*types.Block)
```
Comment (zalmen) [inline sync/validation_queue.go]: you need to validate that this is indeed the block that you asked for (that its id matches `bid`)

File: sync/validation_queue.go [hunk 33]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
 
-func (vq *validationQueue) addDependencies(blk *types.BlockHeader, checkDatabase func(id types.BlockID) (*types.Block, error)) bool {
+func (vq *blockQueue) addDependencies(jobId interface{}, blks []types.BlockID, finishCallback func(res bool) error) (bool, error) {
+	vq.Lock()
+	vq.callbacks[jobId] = finishCallback
 	dependencys := make(map[types.BlockID]struct{})
-	for _, id := range blk.ViewEdges {
+	idsToPush := make([]types.Hash32, 0, len(blks))
+	for _, id := range blks {
 		if vq.inQueue(id) {
```
Comment (zalmen) [inline sync/validation_queue.go]: `InQueue` will return true id `id` is in the visited list. Is it possible that `id` was already visited and all of its dependencies were already notified and hence `jobId` will never get updated that this dep is completed?

File: sync/validation_queue.go [hunk 34]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
-	return len(dependencys) > 0
-}
-
-func (vq *validationQueue) addToDatabase(id types.BlockID) error {
-	if callback, ok := vq.callbacks[id]; ok {
-		if err := callback(); err != nil {
-			return err
-		}
+	for _, id := range idsToPush {
+		vq.addToPending([]types.Hash32{id})
```
Comment (zalmen) [inline sync/validation_queue.go]: if `addToPending` receives a slice of ids, why do you pass them one by one?

File: sync/validation_queue.go [hunk 35]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
-	return nil
-}
 
-func (vq *validationQueue) getMissingBlocks() []types.BlockID {
-	missingBlocks := make([]types.BlockID, 0, len(vq.reverseDepMap))
-	for k := range vq.reverseDepMap {
-		missingBlocks = append(missingBlocks, k)
+	//todo better this is a little hacky
+	if len(dependencys) == 0 {
+		return false, finishCallback(true)
```
Comment (zalmen) [inline sync/validation_queue.go:221]: IMO it is better that you'll handle the error that might be returned by `finishCallback(true)` here. This way `addDependencies` won't need to return any error and it will simplify the caller of this method (which currently need to "remember" that this error actually means that the block had no dependencies and that the execution of its callback failed.

File: sync/validation_queue.go (hunk 33 above)
Comment (zalmen) [reply sync/validation_queue.go]: Or, on second thought, it is better if you'll just return the number of dependencies and let the caller call the callback and handle the case of failure (IMO it is a bit weird that `addDependencies` calls the callback)

File: sync/validation_queue.go [hunk 36]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
+		if len(vq.depMap[dep]) == 0 {
+			delete(vq.depMap, dep)
+			vq.Info("run callback for %v, %v", dep, reflect.TypeOf(dep))
+			if callback, ok := vq.callbacks[dep]; ok {
+				if err := callback(valid); err != nil {
+					vq.Error(" %v callback Failed", dep)
+					continue
+				}
+				delete(vq.callbacks, dep)
+				switch id := dep.(type) {
```
Comment (zalmen) [inline sync/validation_queue.go:179]: Why is it needed? Can it be anything else than a block id?

File: sync/validation_queue.go [hunk 37]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
+	delete(vq.visited, block)
+
+	for _, dep := range vq.reverseDepMap[block] {
+		delete(vq.depMap[dep], block)
+		if len(vq.depMap[dep]) == 0 {
+			delete(vq.depMap, dep)
+			vq.Info("run callback for %v, %v", dep, reflect.TypeOf(dep))
+			if callback, ok := vq.callbacks[dep]; ok {
+				if err := callback(valid); err != nil {
+					vq.Error(" %v callback Failed", dep)
```
Comment (zalmen) [inline sync/validation_queue.go:176]: Don't you need to remove the callback from `callbacks` in such case (like you do in the case of success)

File: sync/validation_queue.go [hunk 38]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
+					continue
+				}
+				delete(vq.callbacks, dep)
```
Comment (zalmen) [inline sync/validation_queue.go]: where do you delete `dep` from `visited`? Why don't you just do a recursive call to `removeFromDepMap` with `dep`?

File: sync/validation_queue.go [hunk 39]
```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
...
 	for {
 		if len(doneQueue) == 0 {
-			return blocks
+			break
 		}
 		block = doneQueue[0]
 		doneQueue = doneQueue[1:]
-		blocks = append(blocks, block)
-		doneQueue = vq.removefromDepMaps(block, doneQueue)
+		doneBlocks = append(doneBlocks, block)
```
Comment (zalmen) [inline sync/validation_queue.go]: `doneBlocks` isn't used anywhere

File: sync/validation_queue.go (hunk 38 above)
Comment (zalmen) [reply sync/validation_queue.go]: Only now I saw that you're calling `removeFromDepMap` in th caller method. So why do you need to remove `dep` from any of the maps if eventually you'll call `removeFromDepMap`?

File: common/types/activation.go (hunk 11 above)
Comment (zalmen) [reply common/types/activation.go]: This comment wasn't addressed yet

File: sync/fetch_queue.go (hunk 20 above)
Comment (zalmen) [reply sync/fetch_queue.go:121]: I saw that in some cases the caller if `Handle` calles `len` and `range` on the resulted map. I thought that it is a problem in case the map is `nil` but I was wrong. I do think that an empty result is nicer than a nil result, but however you prefer

File: sync/validation_queue.go (hunk 31 above)
Comment (almogdepaz) [reply sync/validation_queue.go]: commented out by accident

File: sync/validation_queue.go (hunk 32 above)
Comment (almogdepaz) [reply sync/validation_queue.go]: its done inside newFetchReqFactory with validateItemIds

File: sync/validation_queue.go (hunk 36 above)
Comment (almogdepaz) [reply sync/validation_queue.go:179]: yes, there is a case where the dependency is not a block but a layer, happens when the sync passes the ids of some layer
i can explain f2f

File: sync/validation_queue.go (hunk 35 above)
Comment (almogdepaz) [reply sync/validation_queue.go:221]: its called from the sync for a layer and the validation queue each needs to know the error because they handle it differently

File: sync/validation_queue.go (hunk 34 above)
Comment (almogdepaz) [reply sync/validation_queue.go]: i dident want to add block batch fetching as part of this pr
this is why i do  " block := bjb.items[0].(*types.Block)"
in handleBlock
if you dont mind the clatter i can do it in this pr
theres a task for this that bark opened

File: sync/validation_queue.go (hunk 38 above)
Comment (almogdepaz) [reply sync/validation_queue.go]: first blocks also needs his items cleand, i moved it to updateDependencies before the loop with removeFromDepMap now its clearer

File: sync/validation_queue.go (hunk 33 above)
Comment (almogdepaz) [reply sync/validation_queue.go]: where do we call the callback in addDependencies ?

File: sync/syncer.go (hunk 28 above)
Comment (almogdepaz) [reply sync/syncer.go:291]: faster then validating all the view ...
better suggestions ?

File: sync/requests.go (hunk 27 above)
Comment (almogdepaz) [reply sync/requests.go:198]: i dont see where this is duplicated

File: sync/requests.go (hunk 26 above)
Comment (almogdepaz) [reply sync/requests.go:29]: newFetchReqFactory dose this for all fetch items

File: sync/handler.go (hunk 24 above)
Comment (almogdepaz) [reply sync/handler.go]: this is the same behavior as we have in txs/atxs it needs to change but in a different pr
i changed the return to continue, we still need to to work on the logic in the receiving end in #1401

File: sync/requests.go (hunk 27 above)
Comment (zalmen) [reply sync/requests.go:198]: amm I can't either, I was probably wrong

File: sync/validation_queue.go (hunk 34 above)
Comment (zalmen) [reply sync/validation_queue.go]: ok, up to you to decide

Comment (almogdepaz) [reply sync/validation_queue.go]: changed it we now fetch blocks by batch

File: sync/validation_queue.go (hunk 33 above)
Comment (zalmen) [reply sync/validation_queue.go]: line 213. In case `len(dependencys) == 0`

**Review by zalmen — APPROVED** [review]
